# 인공지능과 가위바위보 하기

## 0.사전준비

- 학습시킬 데이터인 가위, 바위, 보 제스쳐 사진을 웹캠 캡쳐로 준비

- 조원들이 촬영한 사진들을 모아서 압축
- LMS 클라우드 환경에 문제가 있어, 코랩으로 진행할 예정 > 구글 드라이브 마운트 후 원본데이터 업로드


## 1.데이터 전처리

- LMS의 기본 진행에는 없지만, 조원들과 수합한 데이터를 먼저 학습용/테스트용으로 분류하고 싶었음

- 먼저 데이터 파일들의 이름을 일괄적으로 수정하고 싶어서 방법을 찾아보았음 > os모듈을 활용하면 될 것 같다!

In [1]:
import os                    # 모듈 임포트
import numpy as np
import matplotlib.pyplot as plt
import glob
from PIL import Image

In [2]:
# 기본 경로를 path 변수에 담아 주었다.
path = '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/'

### 이왕이면 압축 풀기도 코드로 해보고 싶었다.

- 터미널에서 사용하는 리눅스 명령어들은 파이썬 문법과 함께 사용하기 어려운 것을 느낌.

- 각 디렉터리로 직접 이동하여 압축 풀어주기.

- 파일이 너무 많아 출력 내용은 제거해주었다.

- ! 기로호 시작하는 터미널 명령어는 여러 줄 실행이 안 되는데 %기호를 활용하면 한 셀 안에서 순차적으로 실행이 가능하다고 한다.

In [ ]:
%cd '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/rock'
!unzip '*.zip' # 각 디렉터리에는 하나의 압축 파일만 있어서 편의상 모든 zip파일의 압축을 풀도록 명령

In [ ]:
%cd '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/scissor'
!unzip '*.zip'

In [ ]:
%cd '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/paper'
!unzip '*.zip'

In [3]:
%cd '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/'
!pwd #혹시 모르니 기본 경로로 이동시켜둠


/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper
/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper


### 파일명 바꾸기

In [4]:
os.listdir(path) #기본 경로의 파일명을 리스트로 가져옴

['rock', 'scissor', 'paper', 'rock_scissor_paper.ipynb']

In [5]:
from glob import glob # LMS에서 진행할 때 파일이 자꾸 폴더로 바뀌는(?) 현상을 겪어 glob를 같이 사용해보기로 함
from os.path import isdir

In [6]:
glob(path+'rock/*.jpg')[0] #.jpg로 끝나는 파일의 리스트 중 첫 번째 파일을 확인해봄. 경로까지 다 나온다.

'/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/rock/rock_1578.jpg'

In [7]:
isdir(path) #isdir의 작동도 확인. 대상이 디렉터리면 트루

True

In [8]:
isdir(glob(path+'rock/*.jpg')[0]) #아니면 펄스를 출력

False

In [9]:
def ol(a):
    return os.listdir(a) #경로 호출을 자주 할 것 같아 간단한 함수 생성

ol(path) #잘 작동하는 것 같다

['rock', 'scissor', 'paper', 'rock_scissor_paper.ipynb']

In [73]:
for i in ol(path):
    if i == 'rock' or i == 'scissor' or i == 'paper':    #반환된 리스트 값을 조건으로 지정
        glob_i = glob(path+i+'/*.jpg')                   #통과한 경우 변수 지정        
        k=0
        for j in ol(path+i):                 #각 디렉터리의 jpg파일 수 만큼을 레인지로 지정한 반복문 생성
                        
            name, ext = os.path.splitext(j)       # splitext메서드로 확장자를 분리할 수 있다. 확장자가 포함된 인자를 넣을 경우 확장자와 나머지로 분리한다.

            if ext != '.jpg':
                print(ext)
                pass
                            
            else:
                name = i + '_' + str(k+1)                # name을 rock_1과 같은 모양으로 저장
                new_name = name + ext                  #새롭게 지정될 이름은 name+.jpg의 형태일 것
                os.rename(glob_i[k], os.path.join(path+i, new_name))     
                k += 1           



.zip
.zip
.zip


In [10]:
import sys
print(sys.version)

3.7.12 (default, Sep 10 2021, 00:21:48) 
[GCC 7.5.0]


In [11]:
print(len(ol(path+'rock')))
print(len(ol(path+'scissor')))
print(len(ol(path+'paper')))
print(len(ol(path+'rock'))+len(ol(path+'scissor'))+len(ol(path+'paper')))


2577
2749
2535
7861


In [12]:
print(ol(path+'rock')[:10])
print(ol(path+'scissor')[:10])
print(ol(path+'paper')[:10])
#잘 작동했..나..?

['rock_1578.jpg', 'rock_1579.jpg', 'rock_1580.jpg', 'rock_1582.jpg', 'rock_1581.jpg', 'rock_1583.jpg', 'rock_1584.jpg', 'rock_1585.jpg', 'rock_1586.jpg', 'rock_1587.jpg']
['scissor_1750.jpg', 'scissor_1751.jpg', 'scissor_1752.jpg', 'scissor_1753.jpg', 'scissor_1754.jpg', 'scissor_1755.jpg', 'scissor_1756.jpg', 'scissor_1758.jpg', 'scissor_1757.jpg', 'scissor_1760.jpg']
['paper_1536.jpg', 'paper_1537.jpg', 'paper_1538.jpg', 'paper_1539.jpg', 'paper_1540.jpg', 'paper_1541.jpg', 'paper_1543.jpg', 'paper_1542.jpg', 'paper_1544.jpg', 'paper_1545.jpg']


In [77]:
#예외가 있는지, 새 이름 양식 문자열이 포함되지 않은 파일을 출력해보자

for i in ol(path):
    if i == 'rock' or i == 'scissor' or i == 'paper':
        for j in range(len(ol(path+i))):
            if i in ol(path+i)[j]:
                pass
            else:
                print(ol(path+i)[j])


# 결과가 없는 것을 보니 잘 작동한 것 같다. 클라우드에서 직접 폴더를 훑어본 결과 성공.

### Resize

- 1-7의 코드를 가져옴
- glob.glob가 작동하지 않아서 위에서 처럼 glob메서드를 한 번만 사용하는 것으로 수정
- 구글드라이브는 경로에 root를 넣으니 작동하지 않아서 수정
- 이후 계속 에러가 뜨는 상태였는데, 리네임 코드를 여러번 돌리던 과정에서 문제가 있었던 것으로 추정된다..ㅠ

In [78]:
from PIL import Image

def resize_images(img_path):                                    # 리사이즈 함수 생성
    images=glob(img_path + '/*.jpg')                            # glob으로 경로를 images 변수에 담기
    print(len(images), 'images to be resized.')                 # 이미지 파일의 갯수와 문구 출력
    target_size=(28,28)                                         # 28x28로 사이즈 변환
    for img in images: #[1:]:
        """에러가 뜨고 엄청 고생한 뒤에 보니 압축 파일 확장자가 jpg로 리네임 되어 있었다. 
        리네임 코드에 분명 글롭이 선행되어 있어서 zip파일이 리네임 된 것이 이상한데, 
        코드를 여러번 수정하며 돌리던 과정에서 발생한 사고 같다.
        우선 첫 번째 항목을 무시하고 실행시키기로 결정"""
    
        old_img=Image.open(img)                                 # 원본 이미지를 가져와 새 이미지로 바꾸고, 저장하는 반복문
        new_img=old_img.resize(target_size,Image.ANTIALIAS)
        new_img.save(img, 'JPEG')
    
    print(len(images), 'images resized.')

In [79]:
image_dir_path = '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/rock'

resize_images(image_dir_path)
        

2577 images to be resized.
2577 images resized.


In [80]:
image_dir_path_1 = '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/paper'

resize_images(image_dir_path_1)

2535 images to be resized.
2535 images resized.


In [81]:
image_dir_path_2 = '/content/drive/MyDrive/Aiffel/Exploration/rock_scissor_paper/scissor'

resize_images(image_dir_path_2)

2749 images to be resized.
2749 images resized.


### 분류, 분할, 라벨링

- 전체 데이터를 먼저 모았기 때문에 학습용, 테스트용 데이터 분리를 먼저 하는 것이 좋겠다고 생각되었다.

- 순서상 나중에 배운 사이킷런에는 split메서드가 
있었는데, 이번 익스플로레이션에서 사용할 tensorflow API에서 제공하는 관련 기능은 찾지 못했다. 전체 데이터 갯수를 알고 있기 때문에 random모듈을 사용해 섞은 다음 8:2 정도의 비율로 나눠볼 것.

- 예제로 주어진 load_data 함수를 살펴보면서, 함수 안에서 직접 path를 지정해주기보다 path를 찾아가서 라벨링 하는 것이 더 좋지 않을까 하는 생각이 들어 수정을 시도해 볼 것이다.
라고 했지만 완전히 멸망했다. 아래 코드를 수정하며 대부분의 시간을 날려버렸다!!

In [13]:
import numpy as np
from glob import iglob
from PIL import Image

In [ ]:
'''
def load_data(dir_path):
    img_size=28
    color=3
    number_of_data = 0
    imgs=np.zeros(number_of_data*img_size*img_size*color,dtype=np.int32).reshape(number_of_data,img_size,img_size,color)
    labels=np.zeros(number_of_data,dtype=np.int32) 

    idx=0    
    for img_dir in ol(dir_path):               
        if isdir(path+img_dir) == True:
            number_of_data = len(ol(path+img_dir))            
            print(number_of_data)
            imgs=np.zeros(number_of_data*img_size*img_size*color,dtype=np.int32).reshape(number_of_data,img_size,img_size,color)
            labels=np.zeros(number_of_data,dtype=np.int32) 

            for file in glob(dir_path+img_dir+'/*.jpg'):
                img = np.array(Image.open(file), dtype=np.int32)
                imgs[idx,:,:,:]=img
                labels[idx]=ol(dir_path).index(img_dir)
                idx=idx+1                   
        else:
            pass

    return imgs, labels
'''

In [17]:

def load_data(img_path, number_of_data=7861):  # 가위바위보 이미지 개수 총합에 주의하세요.
    # 가위 : 0, 바위 : 1, 보 : 2
    img_size=28
    color=3
    #이미지 데이터와 라벨(가위 : 0, 바위 : 1, 보 : 2) 데이터를 담을 행렬(matrix) 영역을 생성합니다.
    imgs=np.zeros(number_of_data*img_size*img_size*color,dtype=np.int32).reshape(number_of_data,img_size,img_size,color)
    labels=np.zeros(number_of_data,dtype=np.int32)

    idx=0
    for file in iglob(img_path+'scissor/*.jpg'):
        img = np.array(Image.open(file),dtype=np.int32)
        imgs[idx,:,:,:]=img    # 데이터 영역에 이미지 행렬을 복사
        labels[idx]=0   # 가위 : 0
        idx=idx+1

    for file in iglob(img_path+'rock/*.jpg'):
        img = np.array(Image.open(file),dtype=np.int32)
        imgs[idx,:,:,:]=img    # 데이터 영역에 이미지 행렬을 복사
        labels[idx]=1   # 바위 : 1
        idx=idx+1  
    
    for file in iglob(img_path+'paper/*.jpg'):
        img = np.array(Image.open(file),dtype=np.int32)
        imgs[idx,:,:,:]=img    # 데이터 영역에 이미지 행렬을 복사
        labels[idx]=2   # 보 : 2
        idx=idx+1
        
    print("이미지 개수는", idx,"입니다.")
    return imgs, labels

(img_data, lable) = load_data(path)
img_data_norm = img_data/255.0   # 입력은 0~1 사이의 값으로 정규화

print("img_data: {}".format(img_data.shape))
print("lable: {}".format(lable.shape))

이미지 개수는 7861 입니다.
img_data: (7861, 28, 28, 3)
lable: (7861,)


In [37]:
type(img_data)

numpy.ndarray

In [45]:
def choice_train_test_split(X, y, test_size=0.2, shuffle=True, random_num):
    
    test_num = int(X.shape[0] * test_size)
    train_num = X.shape[0] - test_num
    
    if shuffle:
        np.random.seed(random_state = random_num)
        train_idx = np.random.choice(X.shape[0], train_num, replace=False)

        test_idx = np.setdiff1d(range(X.shape[0]), train_idx)
        
        X_train = X[train_idx,:]
        X_test = X[test_idx]
        y_train = y[train_idx,]
        y_test = y[test_idx]    

    else:
        X_train = X[:train_num]
        X_test = X[train_num:]
        y_train = y[:train_num]
        y_test = y[train_num:]

    return X_train, X_test, y_train, y_test

# 출처: https://rfriend.tistory.com/519 [R, Python 분석과 프로그래밍의 친구 (by R Friend)]

In [46]:
X_train, X_test, y_train, y_test = choice_train_test_split(img_data, lable, 123)

In [47]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(6289, 28, 28, 3) (1572, 28, 28, 3) (6289,) (1572,)


In [36]:
X_test.shape

(6289, 28, 28, 3)

## 2.데이터 학습 및 분석

- 딥러닝 네트워크 설계 따라해보기
- 학습(train)
- 분석(test)

### 학습에 필요한 텐서플로우 라이브러리 불러오기

In [48]:
#!pip install tensorflow==2.4.1

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, LSTM, BatchNormalization
from tensorflow.keras.callbacks import TensorBoard


In [49]:
# 과정 중 발생한 에러가 LMS와 버전 차이 때문인지 확인하기 위해 작성했었다.
print(tf.__version__)
print(keras.__version__)

2.7.0
2.7.0


### 학습 모델 카피

In [53]:
# 익스 1-6의 코드를 카피. 
# 세부적인 내용이 궁금하지만.. 
#여기서 더 혼자 헤메었다가는 업로드 자체를 못 할 것 같아서 다음 기회로.

n_channel_1=16
n_channel_2=32
n_dense=64
n_train_epoch=20

model=keras.models.Sequential()
model.add(keras.layers.Conv2D(n_channel_1, (3,3), activation='relu', input_shape=(28,28,3))) #input shape를 데이터와 맞춰준다.
model.add(keras.layers.MaxPool2D(2,2))
model.add(keras.layers.Conv2D(n_channel_2, (3,3), activation='relu'))
model.add(keras.layers.MaxPooling2D((2,2)))
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(n_dense, activation='relu'))
model.add(keras.layers.Dense(3, activation='softmax')) #최종 분류 할 class의 갯수를 수정

model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 16)        448       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 16)       0         
 2D)                                                             
                                                                 
 conv2d_3 (Conv2D)           (None, 11, 11, 32)        4640      
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 32)         0         
 2D)                                                             
                                                                 
 flatten_1 (Flatten)         (None, 800)               0         
                                                                 
 dense_2 (Dense)             (None, 64)               

### 학습 시키기
- 컴파일 인자들의 내용이 궁금하면서도 앞으로가 조금 두렵다(..)

In [51]:
model.compile(optimizer='adam',
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy'])

X_train_norm = X_train/255.0
model.fit(X_train_norm, y_train, epochs=20)

Epoch 1/20
197/197 [==============================] - 4s 19ms/step - loss: 0.8302 - accuracy: 0.6093
Epoch 2/20
197/197 [==============================] - 4s 19ms/step - loss: 0.3660 - accuracy: 0.8596
Epoch 3/20
197/197 [==============================] - 4s 19ms/step - loss: 0.2040 - accuracy: 0.9237
Epoch 4/20
197/197 [==============================] - 4s 19ms/step - loss: 0.1277 - accuracy: 0.9558
Epoch 5/20
197/197 [==============================] - 4s 19ms/step - loss: 0.0765 - accuracy: 0.9774
Epoch 6/20
197/197 [==============================] - 4s 19ms/step - loss: 0.0510 - accuracy: 0.9863
Epoch 7/20
197/197 [==============================] - 4s 19ms/step - loss: 0.0286 - accuracy: 0.9951
Epoch 8/20
197/197 [==============================] - 4s 19ms/step - loss: 0.0248 - accuracy: 0.9944
Epoch 9/20
197/197 [==============================] - 4s 19ms/step - loss: 0.0204 - accuracy: 0.9948
Epoch 10/20
197/197 [==============================] - 4s 19ms/step - loss: 0.0094 - accura

### 테스트

In [52]:
X_test_norm=X_test/255.0
test_loss, test_accuracy = model.evaluate(X_test_norm, y_test, verbose=2)
print("test_loss: {} ".format(test_loss))
print("test_accuracy: {}".format(test_accuracy))

50/50 - 0s - loss: 0.0088 - accuracy: 0.9994 - 486ms/epoch - 10ms/step
test_loss: 0.008769562467932701 
test_accuracy: 0.9993638396263123


## 3.결과


### 요약
학습조건:
```
n_channel_1=16
n_channel_2=32
n_dense=64
n_train_epoch=20
split의 random state = 123
```

train:
- 12epoch 에서 정확도 0.9998
- 17에서 1.0

test:
- test_loss: 0.008769562467932701 
- test_accuracy: 0.9993638396263123

### 평가(?)
- 테스트 데이터셋을 전제 데이터에서 랜덤스플릿 했고(사실상 벨리데이션 데이터?), 당연히 유사한 이미지가 많아서 정확도가 높았을 것이다.
- 테스트 정확도가 충분히 높아서 더 실행하지는 않았지만 train의 17epoch 부터 쭉 정확도가 1.0이 나와서 과적합이 걱정되었다. 만약 결과가 좋지 않았다면 우선 에폭 수를 줄여 보았을 것 같다.

## 4. 마치며


### 기억에 남는 학습 내용

- rename 메서드는 별로 안정적이지 못하다(..)

- 전처리 과정에서 numpy를 이용해 이미지를 배열로 바꿨는데, 제대로 이해하지는 못했지만 컴퓨터가 읽을 수 있는 형태로 바뀐 것을 확인한 것이 인상적이었다.

- 정말 많이 검색했고 정말 많은 정보를 만났지만 무작정 신뢰해서는 안된다는 큰 교훈을 얻었다.


> split 할 때 함수에 범위가 잘못된 부분이 있어서 학습이 진행되지 않았고 퍼실님 도움을 받아 겨우 문제점을 찾았다.

<br>

### 어려웠던 점

- 사실상 전부 다..

- LMS 클라우드 환경이 제대로 동작하지 않았던 것이 사실 가장 괴로웠다. 

- 수정 하며 여러번 돌려 보기 어려운 코드들이 시간을 엄청 소모시켰다. 

 - os 모듈과 데이터를 직접 수정하는 부분

 - for문 에러가 발생했을 시 이미 값의 일부가 수정된 경우

- 함수를 내부에서 변수를 선언한 경우 전역 변수가 아니어서(?) 리턴을 받는 단계에서 문제가 생기는 경우

<br>

### 다음엔 시도 해보고 싶은 부분
- 특정 샘플을 뽑아서 모델이 제대로 작동하는지 시각적으로 확인해보기

- 데이터를 reshape하고 lableing(?)할 때 함수 안에 디렉터리를 직접 입력하지 않고 리스트를 통해 자동화 하는 시도가 실패했는데 동작시켜보고 싶다.

<br>

### 느낀점 총평

: LMS 클라우드 환경이 내 시스템에서 제대로 동작하지 않는 문제가 있어서 많이 어려움을 겪었다. 아직 솔루션을 받지 못했는데 뾰족한 수가 없다면 포맷이라도 해야 할 것 같아서 시무룩함.

적당한 어려움(?)을 겪을 때에는 주석이나 마크다운으로 메모하기가 편했지만 계속해서 오류를 수정하다 보니 과정을 기억하기 어려웠다. 문제가 발생했을 때 잘 진단하는 것이 곧 실력인 것 같다고 느꼈음.

지금 단계에서는 데이터를 다루는 것 보다 코드를 작성하고 동작시키는 것이  고통스럽지만(?) 더 재미있었다. 

